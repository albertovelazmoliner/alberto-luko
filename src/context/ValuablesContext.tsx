import React, {
  createContext,
  Dispatch,
  SetStateAction,
  useState,
} from "react";
import { Valuable } from "../domain";

type IValuablesContext = {
  valuables: Valuable[];
  setValuables: Dispatch<SetStateAction<Valuable[]>>;
}

const ValuablesContext = createContext<IValuablesContext>({} as IValuablesContext);

export function ValuablesContextProvider(props: { children: React.ReactNode }) {
  const [valuables, setValuables] = useState<Valuable[]>([]);

  return (
    <ValuablesContext.Provider value={{ valuables, setValuables }}>
      {props.children}
    </ValuablesContext.Provider>
  );
}

export default ValuablesContext;
