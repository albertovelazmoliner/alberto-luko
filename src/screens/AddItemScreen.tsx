import { Alert, StyleSheet, View } from "react-native";
import { ValuableForm } from "../components/valuableForm";

import Button from "../components/Button";
import { RootTabScreenProps } from "../navigation/types";
import { colors } from "../theme/colors";
import { useValuables } from "../hooks";
import { useEffect, useState } from "react";

export default function AddItemScreen({
  navigation,
}: RootTabScreenProps<"AddItemScreen">) {
  const { addValuable } = useValuables();
  const [name, setName] = useState("");
  const [description, setDescription] = useState("");
  const [image, setImage] = useState("");
  const [price, setPrice] = useState("");

  const [addButtonDisabled, setAddButtonDisabled] = useState(true);

  useEffect(() => {
    if (name.length > 0 && image.length > 0 && price !== null) {
      setAddButtonDisabled(false);
    } else {
      setAddButtonDisabled(true);
    }
  }, [name, image, price]);

  const showAlert = (message: string) =>
    Alert.alert(
      "Action failed",
      message,
      [
        {
          text: "Ok",
          style: "default",
        },
      ]
    );

  const addValuableHandler = () => {
    try {
      addValuable(name, price, description, image);
      navigation.goBack();
    } catch (err: any) {
      showAlert(err.message);
    }
  };

  return (
    <View style={styles.container}>
      <View style={styles.buttonsContainer}>
        <Button title="Cancel" onPress={() => navigation.goBack()} />
        <Button
          title="Add"
          disabled={addButtonDisabled}
          onPress={addValuableHandler}
        />
      </View>
      <ValuableForm
        onChangeName={(name: string) => setName(name)}
        onChangeDescription={(description: string) =>
          setDescription(description)
        }
        onChangePhoto={(image: string) => setImage(image)}
        onChangePrice={(price: string) => setPrice(price)}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingHorizontal: 20,
    backgroundColor: colors.background,
    paddingTop: 10,
  },
  buttonsContainer: {
    width: "100%",
    marginTop: 16,
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-between",
    backgroundColor: "transparent",
  },
});
