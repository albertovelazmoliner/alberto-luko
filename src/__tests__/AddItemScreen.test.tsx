import AddItemScreen from "../screens/AddItemScreen";
import renderer from "react-test-renderer";
import React from "react";
import { ValuablesContextProvider } from "../context/ValuablesContext";

const navigation: any = { navigate: jest.fn() };
const route: any = { name: "AddItem" };

const whenStable = async () => {
  await renderer.act(async () => {
    await new Promise((resolve) => {
      setTimeout(resolve, 0);
    });
  });
};

describe("test AddItem Screen is rendered properly", () => {
  it("foo", async () => {
    const screen = renderer.create(
      <ValuablesContextProvider>
        <AddItemScreen navigation={navigation} route={route} />
      </ValuablesContextProvider>
    );

    expect(screen.toJSON()).toMatchSnapshot();

    await whenStable();
  });
});
