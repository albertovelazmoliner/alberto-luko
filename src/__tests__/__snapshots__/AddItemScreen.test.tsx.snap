// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test AddItem Screen is rendered properly foo 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#E5E5E5",
      "flex": 1,
      "paddingHorizontal": 20,
      "paddingTop": 10,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginTop": 16,
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 20,
          "left": 20,
          "right": 20,
          "top": 20,
        }
      }
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "opacity": 1,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#2D50E6",
            "fontSize": 17,
          }
        }
      >
        Cancel
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 20,
          "left": 20,
          "right": 20,
          "top": 20,
        }
      }
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "opacity": 1,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#C0BEB8",
            "fontSize": 17,
          }
        }
      >
        Add
      </Text>
    </View>
  </View>
  <View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignContent": "center",
          "borderColor": "lightgrey",
          "borderRadius": 100,
          "borderStyle": "dashed",
          "borderWidth": 2,
          "height": 140,
          "marginBottom": 10,
          "marginEnd": "auto",
          "marginStart": "auto",
          "marginTop": 10,
          "width": 140,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "column",
            "height": 140,
            "justifyContent": "center",
            "margin": 0,
            "width": 140,
          }
        }
      >
        <View>
          <Text />
        </View>
        <Text>
          Add photo
        </Text>
      </View>
    </View>
    <View>
      <Text
        style={
          Object {
            "fontFamily": "Circular-Bold",
            "fontSize": 12,
            "lineHeight": 24,
          }
        }
      >
        Name
      </Text>
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "backgroundColor": "white",
            "borderColor": "lightgrey",
            "borderRadius": 8,
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 42,
            "padding": 8,
            "width": "100%",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          clearButtonMode="while-editing"
          onChangeText={[Function]}
          placeholder="Bracelet"
          rejectResponderTermination={true}
          style={
            Object {
              "fontSize": 18,
              "width": "100%",
            }
          }
          underlineColorAndroid="transparent"
        />
      </View>
    </View>
    <View>
      <Text
        style={
          Object {
            "fontFamily": "Circular-Bold",
            "fontSize": 12,
            "lineHeight": 24,
          }
        }
      >
        Value
      </Text>
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "backgroundColor": "white",
            "borderColor": "lightgrey",
            "borderRadius": 8,
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 42,
            "padding": 8,
            "width": "100%",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          clearButtonMode="never"
          keyboardType="number-pad"
          onChangeText={[Function]}
          placeholder="700"
          rejectResponderTermination={true}
          style={
            Object {
              "fontSize": 18,
              "width": "100%",
            }
          }
          underlineColorAndroid="transparent"
        />
        <View>
          <Text />
        </View>
      </View>
    </View>
    <View>
      <Text
        style={
          Object {
            "fontFamily": "Circular-Bold",
            "fontSize": 12,
            "lineHeight": 24,
          }
        }
      >
        Description
      </Text>
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "backgroundColor": "white",
            "borderColor": "lightgrey",
            "borderRadius": 8,
            "borderWidth": 1,
            "flexDirection": "column",
            "height": 150,
            "padding": 8,
            "width": "100%",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          multiline={true}
          numberOfLines={6}
          onChangeText={[Function]}
          placeholder="Optional"
          rejectResponderTermination={true}
          style={
            Object {
              "fontSize": 18,
              "height": 100,
              "width": "100%",
            }
          }
          underlineColorAndroid="transparent"
        />
      </View>
    </View>
  </View>
</View>
`;
